File: Game.Application/GameJS/src/utils/specifications/IfSpecification.ts
import { ISpecification } from "./ISpecification";

export class IfSpecification<TSituation, TOutput> implements ISpecification<TSituation, TOutput> {
  private readonly condition: ISpecification<TSituation, boolean>;
  private readonly whenTrue: ISpecification<TSituation, TOutput>;
  private readonly whenFalse: ISpecification<TSituation, TOutput>;
  constructor(condition: ISpecification<TSituation, boolean>, whenTrue: ISpecification<TSituation, TOutput>, whenFalse: ISpecification<TSituation, TOutput>) {
    this.condition = condition;
    this.whenTrue = whenTrue;
    this.whenFalse = whenFalse;
  }
  execute(situation: TSituation): TOutput {
    return this.condition.execute(situation)
      ? this.whenTrue.execute(situation)
      : this.whenFalse.execute(situation);
  }
}

File: Game.Domain/Specifications/IfSpecification.cs
namespace Game.Domain.Specifications
{
    public class IfSpecification<TSituation, TOutput> : ISpecification<TSituation, TOutput>
    {
        private readonly ISpecification<TSituation, bool> condition;
        private readonly ISpecification<TSituation, TOutput> whenTrue;
        private readonly ISpecification<TSituation, TOutput> whenFalse;

        public IfSpecification(ISpecification<TSituation, bool> condition, ISpecification<TSituation, TOutput> whenTrue, ISpecification<TSituation, TOutput> whenFalse)
        {
            this.condition = condition;
            this.whenTrue = whenTrue;
            this.whenFalse = whenFalse;
        }

        public TOutput Execute(TSituation situation)
        {
            return condition.Execute(situation)
                ? whenTrue.Execute(situation)
                : whenFalse.Execute(situation);
        }
    }

}
